Implement AA P3HT (the first GROMOS 53A6 polymer)

ricalessandri opened the pull request:
I implemented AA P3HT with the H termini  as discussed; the itp generation is good. The itp can be generated with:

```
polyply gen_itp -lib gromos_53A6_polymers -name P3HT -seq Hter:1 P3HT:6 Hter:1 -o P3HT_n6_gromos.itp
```

fgrunewald commented on polyply/data/gromos_53A6_polymers/Hter.ff:
```diff
@@ -0,0 +1,7 @@
+;
+[ atoms ]
+;  nr  type  resnr  resid  atom  cgnr  charge    mass    total_charge
+    1    HC    1    P3HT   HTe    1    0.245    1.008 ;  0.245
```
The resname cannot be P3HT because it is not the same residue. You could name it TER or HTer or whatever you want. I think with Martini my convention was calling it 'OHter' i.e. with small ter. But whatever floats your boat.

ricalessandri replied:
Right! Fixed (https://github.com/marrink-lab/polyply_1.0/pull/112/commits/88bd340b9e0896e2da9f021e8586f5d18f25f7d9), I went for Hter.

fgrunewald commented on polyply/data/gromos_53A6_polymers/P3HT.gromos.53A6.itp:
```diff
@@ -0,0 +1,198 @@
+; ;;;;; poly(3-hexyl-thiophene) (P3HT)
```
It's probably better to make this an .ff file. It should work by just changing the extension to .ff. The benefit is that it will be treated as vermouth input file so that makes the library a bit more streamline. Also you can include a .bib file with all the citations and than add a `[ cite ]` directive like shown here: https://github.com/marrink-lab/vermouth-martinize/blob/a98098629ace902940c5e3e9dd3cd6f1ec6c0139/vermouth/data/force_fields/elnedyn/aminoacids.ff#L23

In the cite directive there should be all citations for this force-field. Alternatively you can also have a [cite] directive under the molecule in which case the citation only applies to the molecule.

ricalessandri replied:
Making it a `.ff` file are re-running the above `polyply gen_itp ...` gives this:

```
Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 524, in _treat_block_interaction_atoms
    reference = atom_names[reference]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 239, in parse_section
    return method(self, line, lineno, **kwargs)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 323, in _interactions
    _base_parser(
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 801, in _base_parser
    treated_atoms = _treat_block_interaction_atoms(atoms, context, section)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 529, in _treat_block_interaction_atoms
    raise IOError(msg.format(len(context), context.name,
OSError: There are 25 atoms defined in the block "P3HT". Interaction in section "bonds" cannot refer to atom index 31.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/riccardo/miniconda3/envs/polyply/bin/polyply", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 277, in <module>
    main()
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 273, in main
    args.func(args)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/gen_itp.py", line 53, in gen_itp
    force_field = load_library(args.name, args.lib, args.inpath)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 110, in load_library
    read_ff_from_files(all_files, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 75, in read_ff_from_files
    wrapper(parser, path, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 67, in wrapper
    parser(lines, force_field=force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/ffinput.py", line 1003, in read_ff
    return list(director.parse(iter(lines)))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 110, in parse
    result = self.dispatch(line)(line, lineno)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/vermouth-martinize/vermouth/parser_utils.py", line 241, in parse_section
    raise IOError("Problems parsing line {}. I think it should be a "
OSError: Problems parsing line 67. I think it should be a '['moleculetype', 'bonds']' line, but I can't parse it as such.
```

The crucial line seems to be this one?
`OSError: There are 25 atoms defined in the block "P3HT". Interaction in section "bonds" cannot refer to atom index 31.
`
as indeed the line in the itp is the one containing the atom index 31; this line is supposed to define a bond with the next monomer, so indeed 31 exceeds 25, with 25 being the number of atoms in the monomer. I thought this was the way to do it by looking at, e.g., https://github.com/marrink-lab/polyply_1.0/blob/master/polyply/data/gromos_2016H66_polymers/PEG.gromos.2016.itp or https://github.com/marrink-lab/polyply_1.0/blob/master/polyply/data/martini2_polymers/PS.martini.2.itp.

If the file is `.itp` this doesn't happen and the command successfully generates the topology of the polymer

ricalessandri replied:
Regarding the citation. I tried it out by inserting the `[ citations ]` directive and by placing a `citations.bib` file there as done in the vermouth folder you linked above (neat thing, btw!) and commenting out all the lines with atom numbers > 25 from the `.ff`. If you do that, then the command runs, however, you get a new error:

```
Traceback (most recent call last):
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 74, in read_ff_from_files
    parser = FORCE_FIELD_PARSERS[file_extension]
KeyError: 'bib'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riccardo/miniconda3/envs/polyply/bin/polyply", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 277, in <module>
    main()
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/bin/polyply", line 273, in main
    args.func(args)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/gen_itp.py", line 53, in gen_itp
    force_field = load_library(args.name, args.lib, args.inpath)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 110, in load_library
    read_ff_from_files(all_files, force_field)
  File "/Users/riccardo/ric.alessandri/Repos/polyply/polyply_1.0/polyply/src/load_library.py", line 77, in read_ff_from_files
    raise IOError(
OSError: Cannot parse file with extension bib.
```
Polyply not liking to find files with extension `.bib` in there?

fgrunewald replied:
@ricalessandri the bib reader is quite recent. Did you update your vermouth/martinize version already?

ricalessandri replied:
Yep. I repeated the test again today, and updated the repos as well, just to double-check. I still get both errors (the 25 atoms & bib). I'm using the master/main of vermouth and the master/main of polyply.

fgrunewald replied:
well in that case we have ourselves a good old dependency mess up. I'll investigate but thanks for double checking!

fgrunewald replied:
#123 should fix this issues. I forgot that I didn't added the parers for citations to the polyply parsers and also there were some bugs that I had to resolve. So we'll wait for that fix and then bring through this baby

ricalessandri replied:
Ok. See my feedback on https://github.com/marrink-lab/polyply_1.0/pull/123, maybe there's room for some improvements - and not sure how much of that actually depends on vermouth. After that is merged into master, I will then commit things on this PR accordingly.

fgrunewald reviewed (changes requested):
Few small comments. Can you also add a citation to the GROMOS force-field (i.e. for the nonbond parameters)?

And I think it would be useful to integrate this itp also in the integral tests as we have done before.

ricalessandri commented:
@fgrunewald by the way, should I move the "connection to the next monomer" stuff to the `links`? Because it seems that this is what you're doing in the recent PR https://github.com/marrink-lab/polyply_1.0/pull/119. To be compared to "the older way" where, for example, you had something like this line (which involves an atom in the residue+1) https://github.com/marrink-lab/polyply_1.0/blob/cefa4ffe3ea5c3e769b741e8ca9b7be608206855/polyply/data/gromos_2016H66_polymers/PEG.gromos.2016.itp#L16

fgrunewald commented:
> @fgrunewald by the way, should I move the "connection to the next monomer" stuff to the `links`? Because it seems that this is what you're doing in the recent PR #119. To be compared to "the older way" where, for example, you had something like this line (which involves an atom in the residue+1)
> 
> https://github.com/marrink-lab/polyply_1.0/blob/cefa4ffe3ea5c3e769b741e8ca9b7be608206855/polyply/data/gromos_2016H66_polymers/PEG.gromos.2016.itp#L16

@ricalessandri At the moment you're using the hybrid format for input (i.e. links for terminals and polyply format for the main chain). In principle both is fine, but I'm moving GROMOS to the .ff input purely because it makes the library more clean and easier to maintain in the future. There are some features you cannot get with the polyply input, which we will keep for legacy reasons and also it is way easier for people who just want to implement something quick.

ricalessandri commented:
> @ricalessandri At the moment you're using the hybrid format for input (i.e. links for terminals and polyply format for the main chain). In principle both is fine, but I'm moving GROMOS to the .ff input purely because it makes the library more clean and easier to maintain in the future. There are some features you cannot get with the polyply input, which we will keep for legacy reasons and also it is way easier for people who just want to implement something quick.

Got it. I think it makes sense to switch to the `.ff` format for this one too, I'll do that.

ricalessandri commented:
New format `.ff` done - only have issue https://github.com/marrink-lab/polyply_1.0/issues/126 now.

And a question about the tests, @fgrunewald : it seems that you have implemented them differently for the recent GROMOS_2016H66 update (https://github.com/marrink-lab/polyply_1.0/pull/119) than from what we did [before here](https://github.com/marrink-lab/polyply_1.0/pull/98). Which way should I follow?

fgrunewald commented:
@ricalessandri regarding tests I think it is better to do it the 'new' way as in the PR #119 . But we have to wait for that to go through. So for now just have a ref itp at the ready and I think we can merge it early next week.

Are there any more questions on the implementation?

pckroon commented:
If you want to implement one of those integrative tests early (i.e. before #119) is merged you can rebase this branch on top of that one and go ahead. However, depending on how much work still needs doing in #119 (and on which files) you may need to resolve merge conflicts (probably be re-rebasing).

ricalessandri commented:
> @ricalessandri regarding tests I think it is better to do it the 'new' way as in the PR #119 . But we have to wait for that to go through. So for now just have a ref itp at the ready and I think we can merge it early next week.
> 
> Are there any more questions on the implementation?

Ok, I'll have the ref itp ready and wait for that - just buzz me or something.

No more questions about P3HT, I think it's all set. 
What I could do, though, is also add an alternative version of P3HT which we developed [here](https://doi.org/10.1002/adfm.202004799) - a version which is "QM-optimized", i.e., bonded terms are derived from quantum-chemical calculations, but still uses GROMOS 53A6 nonbonded. To be used if one wants to run QM calculations from snapshots directly taken from MD (either after backmapping and/or directly AA structures). What do you think?

(Will open a PR about OPLS and address https://github.com/marrink-lab/polyply_1.0/issues/122#issuecomment-808877605 soon.)

fgrunewald commented:
@ricalessandri you can now add the ref-itp, check the library_tests.py. Steps are:

1 - make a new directory in 'polyply/tests/test_data/library_tests'  and give it the name of the force-field. If it already exists no need to make a new one. 
2 - make a  directory within that dir with the name of the polymer or more broadly the test-name
3 - within that dir make another dir called 'polyply'
4 - in the polyply dir put a file named command with the command that has to be run in the test; in the same directory also place the reference itp file which has to have the same name in the moleculetype section as the one you generate; all other input files for example .json graph files go in the directory one up
5 - edit the test_lib_files.py file and at line 163 add a new item to the list using the syntax '[ff_name, polymer_name]', where the ff_name is the name of the force-field directory and the polymer name the name of the polymer directory.

ricalessandri commented:
Added the test. I also renamed the corresponding folders to the more compact `gromos53A6`, which is more in line with the more compact data folder naming you have adopted recently.

Some checks didn't pass, though, let me know how I can help with that.
